add docs for credits contract #ntrn-365 (#48)
neutron-org/neutron-docs: docs/neutron/token-generation-event/credits.md

# Credits contract

## Overview
Used as a CW20 like token for distributing (vesting) airdropped NTRN tokens.

## Description

> Basically it's a modified for our purposes CW20 token that holds balances for addresses and can transfer them to others.

Reviewer: > that holds balances

this part seems to be unnecessary because this is the way how basic cw20 tokens work — they are simply a mapping address->balances. given the fact that in the beginning of the sentence you highlight that it's a modified version, I think it's reasonable to focus on modifications we've made, otherwise it's confusing

> can transfer them to others

looks like contract itself is capable of transferring tokens to others transfer. maybe we better mention here that the token transfer is only available for the airdrop contract?

Modifications include:
- `Mint` mints to the airdrop address only and requires untrn tokens sent along;
- `Burn`/`BurnFrom`/`Withdraw` not only burn tokens, but also transfer untrn tokens to addresses in 1:1 proportion to burned cNTRN tokens;

> - Only `Withdraw` is permissionless;

Reviewer: actually I've just noticed that Mint is permissionless as well:
<img width="819" alt="image" src="https://user-images.githubusercontent.com/34917380/222657254-8d84de00-b2f4-4328-91f0-cffe3660315e.png">

please fix it in docs and in credits contract (e.g. in Murad's PR)

Author: Its not permissionless - it has checks in execute_mint, that checks that tokeninfo.minter is the same as sender

Reviewer: oh, interesting!

- `Withdraw` has a vesting mechanism attached (tokens are linearly vested for a given amount of time).

The contract is owned by the Neutron DAO and needs to know airdrop and lockdrop contract addresses on initialization for permission purposes.

> DAO executes `Mint()` with untrn's to mint cNTRN funds in 1:1 proportion and assigns newly minted cNTRN funds to the airdrop address.

Reviewer: > DAO executes `Mint()` with untrn's to mint cNTRN

DAO executes `Mint()` with untrn**s sent along** to mint cNTRN

## Main usage scenarios

### Initialization and Mint
1. DAO initializes contract, specifies token info with minter, airdrop and lockdrop contract addresses.
2. DAO executes `Mint()` along with NTRN's funds, that mints cNTRN's in 1:1 proportion and assigns them to airdrop contract balance.

### Airdrop distribution
1. Airdrop calculates the amount of airdrops for users and uses `Transfer(recipient, amount)` together with `AddVesting(address, amount, start_time, duration)` to transfer funds and set vesting schedule (Always a linear vesting with 0 CLIFF time).
2. There are two ways of burning cNTRN's and sending untrn's to user:

> - users can execute `Withdraw()` to burn cNTRN's and get untrn's according to the vesting schedule. They can look into how much they can withdraw now using `WithdrawableAmount(address)` query. `Withdraw()` can be called only after `config.when_withdrawable` passed (Should be set to lockdrop phase 2 end).

Reviewer: `untrn's` is an incorrect form, should be just `untrn` singular. here and in a couple of other places in the doc. the same about `cNTRN's`. haven't seen ETH or BTC mentions like `send us your ETH's` or something, it's always singular

# Credits contract

## Overview
Used as a CW20 token for distributing (vesting) airdropped tokens.

## Description

Basically it's a modified for our purposes CW20 token that holds balances for addresses and can transfer them to others.
Modifications include:

> - mint mints to the airdrop address only and requires NTRN's sent along

Reviewer 2: ```suggestion
- `Mint` mints to the airdrop address only and requires NTRN's sent along
```

> - specific permissions for some operations

Reviewer 2: what operations, what permissions?

> - burn/burn_from/withdraw not only burns tokens, but also transfers NTRN's to addresses in 1:1 proportion to burned cuntrns

Reviewer 2: ```suggestion
- `Burn`/`BurnFrom`/`Withdraw` not only burns tokens, but also transfers NTRN's to addresses in 1:1 proportion to burned ucNTRNs
```

> - withdraw has a vesting mechanism attached

Reviewer 2: ```suggestion
- `Withdraw` has a vesting mechanism attached (tokens are linearly vested for a given amount of time)
```

> Contract instantiated by DAO and holds airdrop and lockdrop contract addresses for permission purposes.

Reviewer 2: ```suggestion
The contract is owned by the Neutron DAO and needs to know airdrop and lockdrop contract addresses for permission purposes.
```

DAO also executes `Mint()` with NTRN's to create CUNTRN funds in 1:1 proportion and assigns CUNTRN funds to the airdrop address.

## Main usage scenarios

### Initialization and Mint
1. DAO initializes contract
2. DAO executes `UpdateConfig(airdrop_address, lockdrop_address)` of contract to assign config addresses for airdrop and lockdrop contract
3. DAO executes `Mint()` along with NTRN's funds, that mints CUNTRNS in 1:1 proportion and assigns them to airdrop contract balance

### Airdrop distribution

> 1. Airdrop calculates amount of needed airdrops for users and uses `Transfer(recipient, amount)` together with `AddVesting(address, amount, start_time, duration)` to transfer funds and set vesting schedule (Always a linear vesting with 0 CLIFF time).

Reviewer 2: ```suggestion
1. Airdrop calculates the amount of airdrops for users and uses `Transfer(recipient, amount)` together with `AddVesting(address, amount, start_time, duration)` to transfer funds and set vesting schedule (Always a linear vesting with 0 CLIFF time).
```

2. Users that have CUNTRN's on balance, have two ways of getting NTRN's:
- can execute `Withdraw()` to burn CNTRNs and get NTRN's according to the vesting schedule. They can look into how much they can withdraw now using `WithdrawableAmount(address)` query. `Withdraw()` can be called only after `config.when_withdrawable` passed (Should be set to lockdrop phase 2 end).

> - lockdrop can execute `BurnFrom(owner, amount)` to send `amount` of NTRN's to `owner` (This action skips vesting as a reward for lockdrop participation).

Reviewer: you say above that

**Users** that have CUNTRN's on balance, **have two ways** of getting NTRN's

but in the second way of getting NTRNs you describe actions from the lockdrop's contract point of view. could you please elaborate the second way so it contains the whole flow from the user's point of view?
